extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java

```diff
@@ -0,0 +1,148 @@
+package io.quarkus.infinispan.client.runtime.cache;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import javax.annotation.Priority;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.commons.CacheException;
+import org.jboss.logging.Logger;
+
+import io.quarkus.infinispan.client.CacheResult;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.TimeoutException;
+import io.smallrye.mutiny.Uni;
+
+@CacheResult(cacheName = "") // The `cacheName` attribute is @Nonbinding.
+@Interceptor
+@Priority(CacheInterceptor.BASE_PRIORITY + 2)
+public class CacheResultInterceptor extends CacheInterceptor {
+
+    private static final Logger LOGGER = Logger.getLogger(CacheResultInterceptor.class);
+    private static final String INTERCEPTOR_BINDING_ERROR_MSG = "The Quarkus Infinispan Client extension is not working properly (CacheResult interceptor binding retrieval failed), please create a GitHub issue in the Quarkus repository to help the maintainers fix this bug";
+    private final Map<Object, ReentrantLock> synchronousGetLocks = new ConcurrentHashMap<>();
```

Comment:
@mkouba Will that `Map` be shared across all interceptions?

Reply:
@karesti Isn't that `Map` supposed to lock cache keys? What if two different caches are invoked at the same time with the same key? One will be locked waiting for the other while it's not needed, right?

Reply:
> @mkouba Will that `Map` be shared across all interceptions of that interceptor?

Not really. This map will be used for all interceptions of a _particular bean instance_. 

Instances of interceptors are _dependent objects_ of the bean instance they intercept. In other words, whenever a bean instance is created a relevant interceptor instance is created as well and bound to the lifecycle of the intercepted bean.

Reply:
So if I have 2 application scope endpoints I have 2 different interceptors if I use CacheResult annotation. Is that allright ? @mkouba

Reply:
Yeah, you would want it isolated per bean since each bean is tied to a specific cache. Otherwise keys from different caches could return the others values.